### PR TITLE
Move kademlia dialer into separate package

### PR DESCRIPTION
**Pull request by cam-a**
This PR moves the kademlia dialer and its methods into a new package called "kademliaclient".

Why: Circular dependencies in kademlia and trust. We need the ability to dial satellites from the trust package, and we also need to import trust into kademlia to verify vouchers with the trusted satellites list. With the dialer separated into its own package, we don't need to import kademlia into trust

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: pkg/dial/dialer.go [hunk 1]
```diff
@@ -18,20 +19,22 @@ import (
 	"storj.io/storj/pkg/transport"
 )
 
-// Dialer is a kademlia dialer
+var mon = monkit.Package()
+
+// Dialer is a dialer
```
Comment (jenlij) [inline pkg/dial/dialer.go]: 😂love that, can we maybe be a little more descriptive? Not your fault, this is a common occurrence across the codebase.

File: pkg/kademlia/kademlia_test.go [hunk 2]
```diff
@@ -127,30 +149,30 @@ func TestBootstrap(t *testing.T) {
 	ctx := testcontext.New(t)
 	defer ctx.Cleanup()
 
-	bn, s, clean := testNode(ctx, "1", t, []pb.Node{})
+	bn, clean := newTestNode(ctx, "1", t, []pb.Node{})
```
Comment (jenlij) [inline pkg/kademlia/kademlia_test.go]: does bn stand for bootstrap node?

File: pkg/kademlia/kademlia_test.go [hunk 3]
```diff
@@ -35,6 +34,12 @@ const (
 	defaultAlpha = 5
 )
 
+type testNode struct {
```
Comment (jenlij) [inline pkg/kademlia/kademlia_test.go]: why is this named testNode?

File: pkg/dial/dialer.go [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // Copyright (C) 2019 Storj Labs, Inc.
 // See LICENSE for copying information.
 
-package kademlia
+package dial
```
Comment (egonelbre) [inline pkg/dial/dialer.go]: Leave this as a subpackage to `kademlia` because it's kademlia specific.

File: pkg/dial/dialer_test.go [hunk 5]
```diff
@@ -42,7 +42,7 @@ func TestDialer(t *testing.T) {
 	{ // PingNode: storage node pings all other storage nodes
 		self := planet.StorageNodes[0]
 
-		dialer := kademlia.NewDialer(zaptest.NewLogger(t), self.Transport)
+		dialer := dial.NewDialer(zaptest.NewLogger(t), self.Transport)
```
Comment (egonelbre) [inline pkg/dial/dialer_test.go]: notice the repetition in the naming.

File: pkg/kademlia/kademlia.go [hunk 6]
```diff
@@ -141,7 +142,7 @@ func (k *Kademlia) DumpNodes(ctx context.Context) (_ []*pb.Node, err error) {
 
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
-func (k *Kademlia) Bootstrap(ctx context.Context) (err error) {
+func (k *Kademlia) Bootstrap(ctx context.Context, transport transport.Client) (err error) {
```
Comment (egonelbre) [inline pkg/kademlia/kademlia.go]: This probably shouldn't need a separate argument for transport client?

Comment (cam-a) [reply pkg/kademlia/kademlia.go]: With the dialer moved to another package, the transport field on it is not reachable from inside `Bootstrap`
https://github.com/storj/storj/pull/2466/files#diff-f6eb0f0cefab0c21ae61c0e1022a854eL186

File: pkg/kademlia/kademlia_test.go (hunk 2 above)
Comment (cam-a) [reply pkg/kademlia/kademlia_test.go]: yes

File: pkg/kademlia/kademlia_test.go (hunk 3 above)
Comment (cam-a) [reply pkg/kademlia/kademlia_test.go]: because it is returned by the testNode function

Comment (cam-a) [reply pkg/kademlia/kademlia_test.go]: I thought it would be better to make a struct to hold the variables rather than return 4 variables out of testNode

Comment (jenlij) [reply pkg/kademlia/kademlia_test.go]: gotcha

File: pkg/kademlia/kademlia.go [hunk 7]
```diff
@@ -183,7 +184,7 @@ func (k *Kademlia) Bootstrap(ctx context.Context) (err error) {
 			// The way FetchPeerIdentityUnverified does is is to do a basic ping request, which
 			// we have now done. Let's tell all the transport observers now.
 			// TODO: remove the explicit transport observer notification
-			k.dialer.transport.AlertSuccess(ctx, &pb.Node{
+			transport.AlertSuccess(ctx, &pb.Node{
```
Comment (egonelbre) [inline pkg/kademlia/kademlia.go]: Was there a reason dialer isn't alerting itself?

File: pkg/kademlia/dial/dialer.go [hunk 8]
```diff
@@ -1,7 +1,7 @@
 // Copyright (C) 2019 Storj Labs, Inc.
 // See LICENSE for copying information.
 
-package kademlia
+package dial
```
Comment (egonelbre) [inline pkg/kademlia/dial/dialer.go]: Try to see whether you can find a better name for this.

File: pkg/kademlia/kademlia.go (hunk 7 above)
Comment (cam-a) [reply pkg/kademlia/kademlia.go]: Apparently it has something to do with the comment above it, but I don't see why we couldn't just call this inside `FetchPeerIdentityUnverified`. Waiting for @jtolds to have some free time to discuss

Comment (cam-a) [reply pkg/kademlia/kademlia.go]: I added a wrapper to call this from dial package

File: pkg/kademlia/dial/dialer.go (hunk 8 above)
Comment (cam-a) [reply pkg/kademlia/dial/dialer.go]: Any suggestions? Perhaps package `dialer`, but change the name of the variable that the dialer is assigned to, ie 
```
type Foo struct {
    dial *dialer.Dialer
}
```
Perhaps permutations of the word "dial" isn't the best way to go

**Pull request merged**